alcor/services/simulations/magnitudes.py

```diff
@@ -0,0 +1,694 @@
+    return estimate_at(star_mass,
+                       x=(min_mass, max_mass),
+                       y=(min_interest_value, max_interest_value))
+
+
+def get_mass_index(*,
+                   star_mass: float,
+                   mass_grid: np.ndarray) -> int:
+    for row_index in range(mass_grid.size() - 1):
+        if mass_grid[row_index] <= star_mass < mass_grid[row_index + 1]:
+            return row_index
+
+
+def interpolate_by_mass(*,
+                        star_mass: float,
+                        star_cooling_time: float,
+                        min_mass: float,
+                        max_mass: float,
+                        min_mass_index: int,
+                        cooling_time_grid: np.ndarray,
+                        pre_wd_lifetime_grid: np.ndarray,
+                        interest_sequence_grid: np.ndarray,
+                        rows_counts: np.ndarray,
+                        by_logarithm: bool,
+                        one_model: bool = False) -> float:
+    max_mass_index = min_mass_index + 1
+
+    if one_model:
+        extrapolated_interest_value = partial(
+                one_white_dwarfs_estimated_interest_value,
+                star_cooling_time=star_cooling_time,
+                cooling_time_grid=cooling_time_grid,
+                interest_sequence_grid=interest_sequence_grid)
+    elif by_logarithm:
+        extrapolated_interest_value = partial(
+                extrapolated_interest_value_by_log,
+                star_cooling_time=star_cooling_time,
+                cooling_time_grid=cooling_time_grid,
+                pre_wd_lifetime_grid=pre_wd_lifetime_grid,
+                interest_sequence_grid=interest_sequence_grid)
+    else:
+        extrapolated_interest_value = partial(
+                get_extrapolated_interest_value,
+                star_cooling_time=star_cooling_time,
+                cooling_time_grid=cooling_time_grid,
+                pre_wd_lifetime_grid=pre_wd_lifetime_grid,
+                interest_sequence_grid=interest_sequence_grid,
+                by_logarithm=by_logarithm,
+                one_model=one_model)
+
+    if star_cooling_time < cooling_time_grid[min_mass_index, 0]:
+        x1 = extrapolated_interest_value(min_row_index=1,
+                                         mass_index=min_mass_index)
+        case_1 = 1
+    elif star_cooling_time >= cooling_time_grid[min_mass_index,
+                                                rows_counts[min_mass_index]]:
+        rows_count = rows_counts[min_mass_index]
+        x1 = extrapolated_interest_value(min_row_index=rows_count,
+                                         mass_index=min_mass_index)
+        case_1 = 1
+    else:
+        for row_index in range(rows_counts[min_mass_index] - 1):
+            if (cooling_time_grid[min_mass_index, row_index]
+                    <= star_cooling_time
+                    <= cooling_time_grid[min_mass_index, row_index + 1]):
```

Comment:
ну я же писал уже более десяти раз, что если нет ветки `else`, то стоит делать `continue`, меня это прям удручает

Reply:
объяснял, что это как минимум избавляет от вложенности

Reply:
и как максимум позволяет думать чище, не держа в голове условия, которым объекты удовлетворяют, лучше вообще найти имя для этого подкласса объектов и написать утилиту, которая просто проверяет условие и возвращает `True`/`False` (можно эту утилиту даже внутри самой функции объявить, прямо здесь, принимать она должна `row_index`), а тут использовать [`filter`](https://docs.python.org/3/library/functions.html#filter) и уже итерироваться по `filter`-объекту без всяких `if`, это разительно сказывается на читаемости, а значит и на всём

Reply:
скорее всего эта часть буде переписана с использованием `np.searchsorted`. пока не буду переписывать

Reply:
всё же предлагаю переписать, потому что здесь явный баг: если мы ни разу не зайдём в условие, то вылетит `NameError` при попытке обращения к объекту `x_1`, так как он проставляется в этом дереве `if`'ов, это всё несложно делается ведь, вроде
```python
# TODO: find a better name, based on exact condition,
# not on the domain logic
def is_star_cooling_time_in_cell(row_index: int) -> bool:
    return (cooling_time_grid[min_mass_index, row_index]
            <= is_star_cooling_time_in_cell
            <= cooling_time_grid[min_mass_index, row_index + 1])


rows_indexes = filter(is_star_cooling_time_in_cell,
                      # using descending order
                      # because we need last entrance only
                      reversed(range(rows_counts[min_mass_index] - 1)))

try:
    row_index = next(rows_indexes)
except StopIteration:
    err_msg = ('No row index was found '
               'for case with star cooling time inside of grid, '
               'grid seems to be unsorted.')
    raise ValueError(err_msg)
else:
    y_1 = cooling_time_grid[min_mass_index, row_index]
    y_2 = cooling_time_grid[min_mass_index, row_index + 1]
    x_1 = interest_sequence_grid[min_mass_index, row_index]
    x_2 = interest_sequence_grid[min_mass_index, row_index + 1]

# do not depend on row index at all
case_1 = False
```
кода на первый взгляд больше, но он более выразителен в отличие от спагетти с `for`-`if` и включает обработку исключительной ситуации

комментарии и сообщение об ошибке стоит поправить основываясь на твоих познаниях в предметной области, они лишь основаны на моём узком понимании из кода

Reply:
ниже аналогичная ситуация

Reply:
но конечно лучше использовать сторонние и специализированные решения, только предварительно проверить, что исключительная ситуация с отсутствием попадания обрабатывается корректно

Reply:
спасибо, что так расписал. я тут кое-что изменил в коде. можешь посмотреть сейчас. там, кстати, в любом случае `x_1` будет объявлен, ошибки не должно быть.
вообще, интересно, что ты думаешь насчет всей этой функции `interpolate_interest_value`.
одно из самых трешовых мест в этом модуле